base/modulus_remainder.h

```diff
@@ -0,0 +1,58 @@
+#ifndef SLINKY_BUILDER_MODULUS_REMAINDER_H
+#define SLINKY_BUILDER_MODULUS_REMAINDER_H
+
+#include <algorithm>
+#include <cassert>
+#include <cstdint>
+
+
+namespace slinky {
+
+// Based on https://github.com/halide/Halide/blob/main/src/ModulusRemainder.h
+/** The result of modulus_remainder analysis. These represent strided
+ * subsets of the integers. A ModulusRemainder object m represents all
+ * integers x such that there exists y such that x == m.modulus * y +
+ * m.remainder. Note that under this definition a set containing a
+ * single integer (a constant) is represented using a modulus of
+ * zero. These sets can be combined with several mathematical
+ * operators in the obvious way. E.g. m1 + m2 contains (at least) all
+ * integers x1 + x2 such that x1 belongs to m1 and x2 belongs to
+ * m2. These combinations are conservative. If some internal math
+ * would overflow, it defaults to all of the integers (modulus == 1,
+ * remainder == 0). */
+struct modulus_remainder {
+    modulus_remainder() = default;
+    modulus_remainder(int64_t m, int64_t r)
+        : modulus(m), remainder(r) {
+    }
+
+    int64_t modulus = 1, remainder = 0;
```

Comment:
Should this be `index_t`? Or maybe this should be a template? (it might work with `T = expr` too...)

Reply:
Right, I also thought about using index_t, but it's defined in runtime/buffer.h and it seemed wrong to refer it in base/. Maybe we should refactor it?

Reply:
Maybe this really should be `int64_t`. It seems like the simplifier might behave differently on 32-bit vs. 64-bit targets if this is `index_t` (modulus-remainder stuff might overflow at fairly reasonable points due to the multiply), which doesn't seem good. I guess that's already the case though with constant folding and other simplifications related to constants.

Other options:
- Template it (all the prerequisite arithmetic is already templated I think?)
- Just use `std::ptrdiff_t` instead of our own alias.